lightblue/src/main/java/org/esbtools/eventhandler/lightblue/LightblueDocumentEventRepository.java

```diff
@@ -147,69 +203,299 @@ public void markDocumentEventsProcessedOrFailed(
+        supportedAndEnabled.retainAll(canonicalTypesToProcess);
+        return supportedAndEnabled.toArray(new String[supportedAndEnabled.size()]);
     }
 
     /**
-     * Creates {@link LightblueDocumentEvent}s from entities using
-     * {@link #documentEventFactoriesByType}, optimizes away superseded and merge-able events, and
-     * populates {@code entitiesToUpdate} with entity status updates that should be persisted to the
-     * document event entity collection before releasing locks.
-     *
-     * <p>The returned list may include net-new events as the result of merges. These events do not
-     * yet have an associated <em>persisted</em> entity, and therefore have no id's.
+     * Within locked events, inserts new entities and updates existing with status, date, and
+     * survivor id updates.
      *
-     * @param maxEvents The maximum number of parsed document events to return. Not to be confused
-     *                  with {@link #documentEventBatchSize}.
-     * @param documentEventEntities The priority-first entities from lightblue.
-     * @param entitiesToUpdate A mutable collection of entities who's status modifications should be
-     *                         persisted back into lightblue.
-     * @param requester Used by {@link LightblueDocumentEvent} implementations to lookup their
-     *                  corresponding documents.
-     * @return An optimized list of parsed or merged document events ready to be published. No more
-     * than {@code maxEvents} will be returned. List may include events with newly computed entities
-     * that are not yet persisted.
+     * <p>Checks for lost locks before persisting, dropping and logging those lost.
      */
-    private List<LightblueDocumentEvent> parseAndOptimizeDocumentEventEntitiesUpTo(int maxEvents,
-            DocumentEventEntity[] documentEventEntities, List<DocumentEventEntity> entitiesToUpdate,
-            LightblueRequester requester) {
-        List<LightblueDocumentEvent> optimized = new ArrayList<>(maxEvents);
+    private List<LightblueDocumentEvent> persistNewEntitiesAndStatusUpdatesToExisting(
+            LockedResources<SharedIdentityEvents> identityLocks) throws LightblueException {
+        DataBulkRequest insertAndUpdateEvents = new DataBulkRequest();
+
+        List<SharedIdentityEvents> lostLocks = identityLocks.checkForLostResources();
+        List<LightblueDocumentEvent> savedEvents = new ArrayList<>();
+
+        // TODO: We make single request per event here (wrapped in bulk request). Maybe could optimize.
+        // Right now each event may have different processing date which we are looking for.
+        // Could probably change that so processing dates were more grouped.
+        // See: https://github.com/esbtools/event-handler/issues/11
+        for (SharedIdentityEvents lockedEvents : identityLocks.getResources()) {
+            if (lostLocks.contains(lockedEvents)) {
```

Comment:
Unnecessary complexity: there is no need to put lostLocks in a list, then perform a lookup. lockedEvents.isLocked() would be easier.